File: CppSolution/BinaryTreeZigzagLevelOrderTraversal.cpp
#include <iostream>
#include <vector>
#include <queue>
#include "Utility.h"
using namespace std;

class Solution 
{
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) 
    {
    	queue<TreeNode*> tq;
    	queue<int> lq;
    	vector<vector<int> > res;
    	vector<int> tmp;
    	if (root == NULL)
    		return res;
    	tq.push(root);
    	lq.push(1);
    	int level;
    	int lastlevel = 1;
    	TreeNode *p;
    	while (!tq.empty())
    	{
    		level = lq.front();
    		lq.pop();
    		if (level > lastlevel)
    		{
    			res.push_back(tmp);
    			tmp.clear();
    			lastlevel = level;
    		}
    		p = tq.front();
    		tq.pop();
    		tmp.push_back(p->val);
    		if (p->left != NULL)
    		{
    			tq.push(p->left);
    			lq.push(level + 1);
    		}

    		if (p->right != NULL)
    		{
    			tq.push(p->right);
    			lq.push(level + 1);
    		}
    	}	
    	if (res.size() < level)
    		res.push_back(tmp);
    	for (int i = 1; i < res.size(); i += 2)
    	{
    		reverse(res[i].begin(), res[i].end());
    	}
    	return res;
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/Linkedin/BinaryTreeZigzagLevelOrderTraversal.java
package Linkedin;

/**
 * Created by xyunpeng on 2/6/16.
 */
public class BinaryTreeZigzagLevelOrderTraversal {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();

        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        queue.offer(null);
        List<Integer> level = new ArrayList<>();
        boolean reverse = false;
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node == null) {
                if (reverse) {
                    Collections.reverse(level);
                }
                reverse = !reverse;
                result.add(level);
                level = new ArrayList<>();
                if (!queue.isEmpty()) {
                    queue.offer(null);
                }
            } else {
                level.add(node.val);

                if (null != node.left) {
                    queue.offer(node.left);
                }

                if (null != node.right) {
                    queue.offer(node.right);
                }
            }
        }

        return result;
    }
}
